packages/communication/src/utils/tx-params-normalizer.ts

```diff
@@ -0,0 +1,62 @@
+import { NodeCommunicationWrapper } from '..'
+import { CeloTx } from '../commons'
+
+function isEmpty(value: string | undefined) {
+  return (
+    value === undefined ||
+    value === null ||
+    value === '0' ||
+    value.toLowerCase() === '0x' ||
+    value.toLowerCase() === '0x0'
+  )
+}
+
+export class TxParamsNormalizer {
+  private chainId: number | null = null
+  private gatewayFeeRecipient: string | null = null
+
+  constructor(readonly communication: NodeCommunicationWrapper) {}
+
+  public async populate(celoTxParams: CeloTx): Promise<CeloTx> {
+    const txParams = { ...celoTxParams }
+
+    if (txParams.chainId == null) {
+      txParams.chainId = await this.getChainId()
+    }
+
+    if (txParams.nonce == null) {
+      txParams.nonce = await this.communication.nonce(txParams.from!.toString())
+    }
+
+    if (!txParams.gas || isEmpty(txParams.gas.toString())) {
+      txParams.gas = await this.communication.estimateGas(txParams)
+    }
+
+    if (!txParams.gasPrice || isEmpty(txParams.gasPrice.toString())) {
+      txParams.gasPrice = await this.communication.gasPrice(txParams.feeCurrency)
+    }
+
+    return txParams
+  }
+
+  private async getChainId(): Promise<number> {
+    if (this.chainId === null) {
+      this.chainId = await this.communication.chainId()
+    }
+    return this.chainId
+  }
+
+  // @ts-ignore - see comment above
```

Comment:
what comment is this referring to?

Reply:
Hahaha, nice catch. It was a comment removed from master 5 months ago:
```
    Right now, Forno does not expose a node's coinbase so we can't
    set the gatewayFeeRecipient. Once that is fixed, we can reenable
    this.
```
and its ignored because it's not used in the class
I will add again the comment, but this was basically a copy/paste, haha